app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,168 @@
+  rescue_from(ActiveJob::DeserializationError) do |exception|
+    System::ErrorReporting.report_error(exception)
   end
 
-  def perform(_object, _caller_worker_hierarchy = [], _background_destroy_method = 'destroy')
-    build_batch
-  end
+  queue_as :deletion
 
-  def on_success(_, options)
-    on_finish('on_success', options)
-  end
+  # until_executed seems to rely on #after_perform which is skipped on failure so not sure whether retries will work.
+  # additionally if we want to reschedule ourselves in #after_perform, it is not nice to rely on flaky order
+  # see https://github.com/3scale/activejob-uniqueness/blob/main/lib/active_job/uniqueness/strategies/until_executed.rb
+  unique :until_and_while_executing, lock_ttl: 6.hours
 
-  def on_complete(_, options)
-    on_finish('on_complete', options)
-  end
+  # better limit by available  `delete` executors
+  # sidekiq_throttle concurrency: { limit: 10 }
 
-  def on_finish(method_name, options)
-    workers_hierarchy = options['caller_worker_hierarchy']
-    info "Starting DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-    object = GlobalID::Locator.locate(options['object_global_id'])
-    background_destroy_method = @background_destroy_method.presence || 'destroy'
-    DeletePlainObjectWorker.perform_later(object, workers_hierarchy, background_destroy_method)
-    info "Finished DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-  rescue ActiveRecord::RecordNotFound => exception
-    info "DeleteObjectHierarchyWorker##{method_name} raised #{exception.class} with message #{exception.message}"
+  before_perform do |job|
+    info "Starting #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
   end
 
-  protected
+  after_perform do |job|
+    if @remaining_hierarchy.present?
+      self.class.perform_later(*@remaining_hierarchy)
+      msg = " iteration of"
+    end
+    info "Finished#{msg} #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
+  end
 
-  delegate :info, to: 'Rails.logger'
+  class << self
+    # convenience method to schedule deleting an active record object
+    def delete_later(ar_object)
+      perform_later *hierarchy_entries_for(ar_object)
+    end
 
-  attr_reader :object, :caller_worker_hierarchy
+    private
 
-  def build_batch
-    batch = Sidekiq::Batch.new
-    batch.description = batch_description
-    batch_callbacks(batch) { batch.jobs { destroy_and_delete_associations } }
-    batch
-  end
+    # @return the hierarchy entries to handle deletion of an object
+    def hierarchy_entries_for(ar_object)
+      return [] unless ar_object.persisted? # e.g. when calling Proxy#oidc_configuration a new object can be generated
 
-  def batch_description
-    "Deleting #{object.class.name} [##{object.id}]"
-  end
+      if ar_object.is_a?(Account) && !ar_object.should_be_deleted?
```

Comment:
Maybe?

```suggestion
      if ar_object.is_a?(Account) && ar_object.should_not_be_deleted?
```

Also, I'm trying to find the definition of `Account#scheduled_for_deletion?` and couldn't find it, do you know where is it?

Reply:
see https://github.com/3scale/porta/blob/master/app/models/account/states.rb

Reply:
I think `should_be_deleted?` should only be checked for provider, because buyer accounts are not marked for deletion.

Reply:
But buyer accounts inherit should be deleted status from their provider
```
    def will_be_deleted?
      scheduled_for_deletion? || (buyer? && provider_account.try(:should_be_deleted?))
    end
```
So no issue with that check. Except that background deleting a buyer without deleting a provider will not be ever possible.

Should it be possible?